src/autodiff.jl

```diff
@@ -1,16 +1,19 @@
-function autodiff(f!, initial_x::Vector)
+function autodiff(f!::Function, initial_x::AbstractArray)
 
-    permf! = (fx, x) -> f!(x, fx)
+    fvec! = reshape_f(f!, initial_x)
+    permf! = (fx::AbstractVector, x::AbstractVector) -> fvec!(x, fx)
 
-    fx2 = copy(initial_x)
-    jac_cfg = ForwardDiff.JacobianConfig(nothing, initial_x, initial_x)
-    g! = (x, gx) -> ForwardDiff.jacobian!(gx, permf!, fx2, x, jac_cfg)
+    fx2 = vec(copy(initial_x))
```

Comment:
What does `vec` do for  `AbstractArray` types other than `Array`? I realize this is implementation dependent, but is there a clear statement about _what_ it is supposed to do?

Reply:
The default implementation just reshapes arbitrary `AbstractArray` to a one-dimensional vector and returns the unchanged vector for `AbstractVector`, see https://github.com/JuliaLang/julia/blob/2b983d41f064c39c0b73f6c7695cc13e0de45c03/base/abstractarraymath.jl#L38-L39 with some additional documentation on top.

Reply:
It's basically a view that changes the size so that way it will always index linearly.